chalk-parse/src/ast.rs

```diff
@@ -175,30 +302,60 @@ impl PolarizedTraitRef {
     }
 }
 
+/// A valid Rust identifier, along with a `Span` specifying line numbers.
 #[derive(Copy, Clone, Debug)]
 pub struct Identifier {
     pub str: InternedString,
     pub span: Span,
 }
 
+/// A where clause; either a normal Rust trait bound, or a Chalk-specific clause.
+// TODO: Add a "See" that points to a list of Chalk's where clauses, and a
+// description of each.
 pub enum WhereClause {
+    /// A normal rust Trait bound.
+    /// Example: `Foo<T>: Bar`
     Implemented { trait_ref: TraitRef },
+
+    // TODO: Renaming a projection, I think...?
     Normalize { projection: ProjectionTy, ty: Ty },
+
+    /// Assert that a projection is equal to a specific type.
+    /// Example: `<Foo<T> as Bar>::Item == T`
     ProjectionEq { projection: ProjectionTy, ty: Ty },
+
+    // TODO: What does "well formed" mean?
     TyWellFormed { ty: Ty },
+
+    // TODO: What does "well formed" mean?
     TraitRefWellFormed { trait_ref: TraitRef },
+
+    // TODO: What does "from env" mean?
```

Comment:
hmm, Well Formed and FromEnv ought to be explained here

https://rust-lang-nursery.github.io/rustc-guide/traits-implied-bounds.html

but seems like I didn't write it yet =)

Reply:
there is some coverage here: 

https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html#implied-bounds